### PR TITLE
Implement new exercise tree endpoint. Fixes #603

### LINKED ISSUE #603: Implement exercise/chapter tree as an endpoint to the API
# Course exercise API endpoint does not sort the exercises

The API endpoint for listing course exercises (http://localhost:8000/api/v2/courses/1/exercises/) does not seem to sort the output. The exercises should be output in the same order as they are listed in the course hierarchy (see the course pages "course materials" or "exercise results"). Note that the API endpoint outputs course modules (exercise rounds) and lists the module's exercises under it. 

Sorting the course modules is easy based on their `order` fields, but sorting the exercises is hard because learning objects can be nested (typically, there are many chapters in a module and each chapter has child exercises). The A+ web pages use the cache for ouputting the course contents and the sorting is done in the cache class (https://github.com/apluslms/a-plus/blob/d34f94b7c9a8e4d22cd28d08968b36787bb344c4/exercise/cache/hierarchy.py, https://github.com/apluslms/a-plus/blob/d34f94b7c9a8e4d22cd28d08968b36787bb344c4/exercise/cache/content.py). Could the API use the same cache for outputting a sorted list of exercises or should the API implement the sorting in another way?

The API endpoint: https://github.com/apluslms/a-plus/blob/d34f94b7c9a8e4d22cd28d08968b36787bb344c4/course/api/views.py#L63
That line refers to CourseModules, but the exercises are retrieved in the serializer: https://github.com/apluslms/a-plus/blob/d34f94b7c9a8e4d22cd28d08968b36787bb344c4/course/api/full_serializers.py#L44

### PR DESCRIPTION
**Pull request by skulonen**
# Description

**What?**

Add new exercise tree endpoint (`/api/v2/courses/{course_id}/tree`) to A+ API. The new endpoint presents the learning objects of a course in a tree-like hierarchy, unlike the regular `exercises` endpoint, which presents them as a flat list.

**Why?**
The flat exercise list does not sort the exercises in the hierarchical order. This new endpoint is also necessary for the IntelliJ plugin.

**How?**

Created a new viewset (`CourseExerciseTreeViewSet`) which accesses the cached content of the course instance, and two new serializers for outputting the data (`TreeCourseModuleSerializer`, `TreeExerciseSerializer`). The viewset doesn't derive from `ReadOnlyModelViewSet` and the serializers don't inherit from `AplusModelSerializer` because they don't operate on model objects. Instead, they operate on `dict`s that the course instance's cached content contains.

The reason for using the cached content is that it is already in a hierarchical structure.

Fixes #603

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Tested that the exercise tree was returned correctly in cases where there are multiple modules, where the hierarchy is deeply nested, where a module has no exercises, and other cases.

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [ ] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

**Comment by skulonen** [comment]
I asked Olli Kinnunen for feedback regarding the structure of the response, and after discussing it with his team, Olli confirmed that it is satisfactory.

File: course/api/full_serializers.py [hunk 1]
```diff
@@ -219,3 +223,16 @@ class Meta(CourseStudentGroupBriefSerializer.Meta):
             }
         }
 
+
+class TreeCourseModuleSerializer(serializers.Serializer):
+    id = serializers.IntegerField()
```
Comment (Mankro) [inline course/api/full_serializers.py:228]: I wonder if we should include other additional `CourseModule` fields in this serializer.

File: course/api/views.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from rest_framework import filters, generics, permissions, viewsets, status, mixins
+from rest_framework import filters, generics, permissions, viewsets, status, mixins, views
```
Comment (Mankro) [inline course/api/views.py:1]: You added import for `views`, but is it actually used anywhere?

File: course/api/views.py [hunk 3]
```diff
@@ -68,6 +68,20 @@ def get_object(self):
         return self.get_member_object('module', 'Exercise module')
 
 
+class CourseExerciseTreeViewSet(CourseResourceMixin,
+                                viewsets.ViewSet):
+    serializer_class = TreeCourseModuleSerializer
+
+    def list(self, request, *args, **kwargs):
+        serializer = self.serializer_class(
+            self.content.data['modules'],
```
Comment (Mankro) [inline course/api/views.py:77]: Just writing down that `self.content` comes from the super class `CourseResouceMixin - CourseInstanceBaseMixin`. It is the course cache `CachedContent`.

* https://github.com/apluslms/a-plus/blob/46386ac304ebfb8dc00bfd2bffdb7d0d01d6ae8a/course/viewbase.py#L61
* https://github.com/apluslms/a-plus/blob/46386ac304ebfb8dc00bfd2bffdb7d0d01d6ae8a/exercise/cache/content.py#L10

**Review by Mankro — APPROVED** [review]
This looks excellent in a quick overview!

* Does the tree include both exercises and chapters? Looks like yes, just double-checking.
* You could add doc strings to the new classes. It is hard to see from the code that the data comes from the cache. The new serializer classes are very dependent on the structure of the cache object.

File: course/api/views.py (hunk 2 above)
Comment (skulonen) [reply course/api/views.py:1]: It is not, I'll remove it.

File: course/api/full_serializers.py (hunk 1 above)
Comment (skulonen) [reply course/api/full_serializers.py:228]: The IntelliJ team was happy with the fields currently available in the response. But if you think some specific fields should be added, I can do that.

**Comment by skulonen** [comment]
@Mankro I've added the doctrings, removed the unnecessary import and squashed the commits.

>Does the tree include both exercises and chapters? Looks like yes, just double-checking.

Yes. It has to, because exercise can be children of chapters, and things would become really messy if we tried to exclude the chapters but include their child exercises.

**Review by Mankro — APPROVED** [review]

**Pull request merged**